### ISSUE TITLE
JSX ternaries: `null` is collapsed, but `undefined` is not

### ISSUE THREAD

**Issue by suchipi** [prettier/prettier#3552]
**Prettier 1.9.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAggB0wIQIYBOAOlAASkCWkUAShQOYAWMAvMAM4QC2cMjFUeqQD8pVO0x4oAPgA20IfAAeMUnzyqK7UvMGl6EONsZwCcUpzWMNVraQBGZvAGt2qAPQSp00klJQAV1lZAF8ScipoABk4ADNWDm5efj1RcUkZXUU4FSsbOyz9Q2NTc0t1VT47RzgXN08Mnz8AqAATOIE4VrCod2kQABoQCEwYCmh2ZFBCAggAdwAFQgRJlDwANwgKVsGQPHYYZFi8WXY4Icc8MGdeAGVJMAF6ZBgCAPOQVogwI5OzoYEZwIMAWBDw9C4eF+pw+ACt2EocGDrnc8Dwop1of9hgEYJhcQAmLEfSQEIHIED2PD2ACeul2mAIAhgAHVtnxkAAOAAMQ0ZEDOLLBmApjKMpnWcF2ZgAjgEKGZQeDIcShmcuBQXm8PuwnrI4ABFAIQeCqkAwals1ocpAEoavPAUWRPADC3BVKCg0ClQwCZwAKtTVscYSEQkA)

**Input:**
```jsx
<AppBar
  iconRight={something ? <span>long text that is long goes here so that this breaks</span> : null}
  iconLeft={something ? <span>long text that is long goes here so that this breaks</span> : undefined}
/>
```

**Output:**
```jsx
<AppBar
  iconRight={
    something ? (
      <span>long text that is long goes here so that this breaks</span>
    ) : null
  }
  iconLeft={
    something ? (
      <span>long text that is long goes here so that this breaks</span>
    ) : (
      undefined
    )
  }
/>;

```

**Expected behavior:**
```jsx
<AppBar
  iconRight={
    something ? (
      <span>long text that is long goes here so that this breaks</span>
    ) : null
  }
  iconLeft={
    something ? (
      <span>long text that is long goes here so that this breaks</span>
    ) : undefined
  }
/>;

```

When I wrote the JSX ternary handling that special-cased `null`, I didn't also do so for `undefined` because it was an invalid node type at the time. However, since it is a valid node type in React 16, it should probably be treated the same as `null`. Maybe `false` should collapse, too (though it's kinda weird to put in a ternary).

**Comment by duailibe** [comment]
What about numbers or strings? They are valid nodes in React 16 too

**Comment by suchipi** [comment]
Maybe, but I think part of the reason not to collapse null is because it's like "nothing", and it feels weird to give "nothing" its own line with parens and etc. Strings and numbers (and identifiers) are "something" so I feel kinda differently about them.

**Comment by duailibe** [comment]
Yeah an empty string denotes "nothing" though

**Comment by thorn0** [comment]
fixed by #7504
